storage: ensure no filesystem use after close in test builds

jbowens opened the pull request:
In test builds, panic if any of the VFS filesystem facilities are used after the Engine is closed.

Epic: none
Informs #106119.
Resolves #110293.
Release note: None

jbowens committed:
storage: ensure no filesystem use after close in test builds

In test builds, panic if any of the VFS filesystem facilities are used after
the Engine is closed.

Epic: none
Informs #106119.
Release note: None

yuzefovich commented:
Hey @jbowens, should we push this over the finish line? I ran only into a single failure #110293 when I polished #106459 last week (with this PR cherry-picked), so I think it should be almost good to go.

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Nice!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106195#-:-Ne9sRCB9-J5_x8dMUqM:bli1pub)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @itsbilal)



<!-- Sent from Reviewable.io -->

jbowens committed:
kvserver: fix TestCheckConsistencyInconsistent's VFS use

Fix TestCheckConsistencyInconsistent to not access an Engine's filesystem after
closing the Engine. Instead, acquire a vfs.FS directly from the sticky VFS
registry and use that to check for the inconsistency death rattle.

Resolves #110293.
Epic: none
Release note: none

erikgrinaker commented:
Nice, thanks for the test fix! We hadn't gotten around to this yet.

pavelkalinnikov commented on pkg/kv/kvserver/consistency_queue_test.go:
```diff
@@ -337,6 +337,7 @@ func TestCheckConsistencyInconsistent(t *testing.T) {
 
 	// Write some arbitrary data only to store on n2. Inconsistent key "e"!
 	s2 := tc.GetFirstStoreFromServer(t, 1)
+	s2AuxDir := s2.TODOEngine().GetAuxiliaryDir()
```
nit: comment why we can't just inline this below, e.g.:

```golang
// Remember the path to s2 aux directory in storage. We can't use Engine
// after s2 crashes below, so will read from its filesystem directly.
```

pavelkalinnikov commented on pkg/kv/kvserver/consistency_queue_test.go:
```diff
@@ -412,8 +413,8 @@ func TestCheckConsistencyInconsistent(t *testing.T) {
 	// A death rattle should have been written on s2.
-	eng := s2.TODOEngine()
-	f, err := eng.Open(base.PreventedStartupFile(eng.GetAuxiliaryDir()))
+	fs := stickyVFSRegistry.Get(strconv.FormatInt(int64(1), 10))
```
nit: could just inline

```golang
fs := stickyVFSRegistry.Get("1")
```

pav-kv reviewed (commented):
The consistency checker test fix LGTM % nits. Thanks for fixing!

jbowens reviewed (commented):
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106195#-:-NeJbrBI4AzAvP_nBkUG:bz5mair)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal and @pavelkalinnikov)

___
*[`pkg/kv/kvserver/consistency_queue_test.go` line 340 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106195#gh-1326217652:-NeJalX_7AINJdhk1KaB:b-bicsgv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8deacf6d320ae24d8d3dca425210342460436bbf/pkg/kv/kvserver/consistency_queue_test.go#L340)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

nit: comment why we can't just inline this below, e.g.:

```golang
// Remember the path to s2 aux directory in storage. We can't use Engine
// after s2 crashes below, so will read from its filesystem directly.
```
</blockquote></details>

Not sure I follow, but the Store and Engine are stopped down below where we use `s2AuxDir`. 
___
*[`pkg/kv/kvserver/consistency_queue_test.go` line 416 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106195#gh-1326223818:-NeJbk6-DyXepufFK4d0:b2juzuz) ([raw file](https://github.com/cockroachdb/cockroach/blob/8deacf6d320ae24d8d3dca425210342460436bbf/pkg/kv/kvserver/consistency_queue_test.go#L416)):*
<details><summary><i>Previously, pavelkalinnikov (Pavel Kalinnikov) wrote…</i></summary><blockquote>

nit: could just inline

```golang
fs := stickyVFSRegistry.Get("1")
```
</blockquote></details>

Heh, was mindlessly copy-pasting. Done.


<!-- Sent from Reviewable.io -->

jbowens commented:
TFTRs!

bors r+
